### PR TITLE
Added ability to play multiple audio files in web review(and one small bug fix).

### PR DESCRIPTION
**Pull request by jooseef**

### CODE REVIEW

File: mnemosyne/web_server/review_wdgt.py [hunk 1]
```diff
@@ -186,7 +186,7 @@ def to_html(self):
             self.answer = "&nbsp;"
 
         hide_answer_css = ""
-        if card.fact_view.a_on_top_of_q:
+        if None != card and card.fact_view.a_on_top_of_q:
```
Comment (gbear605) [inline mnemosyne/web_server/review_wdgt.py:189]: The idiomatic way to check if something is not None in Python is `card is not None`

Comment (jooseef) [reply mnemosyne/web_server/review_wdgt.py:189]: Okay, thanks.

**Comment by jooseef** [comment]
I wanted to touch the original source code as little as possible. But well, directly inserting some pieces of new code will make inserting JavaScript easier and more efficient. That's good, I like that. Do you plan to release the next version?

**Comment by pbienst** [comment]
There's no timeline for the next version, but if you want, I can make a release when your code is merged.

**Comment by jooseef** [comment]
Ah, ok. I misunderstood. Yes, the next version is almost done. :-)
For me you don't need to make a new release, as I can run my own build. Btw do you use a grapical IDE for development? Which one?

I just have noticed only now, when the question displays in the browser, method "run" in file "simple_html5_audio.py" is called twice.
Why is it called twice, is it necessary?
That makes my coding a bit unhandy now.

**Comment by pbienst** [comment]
Looks a lot better!

But I'm not really understanding when you are incrementing the id. It seems to me that all player ids in the question have id=1, and all those in the answer id=2. Is that what's happening? I would expect that you'd want all the player ids different, even within Q and within A.

BTW, why don't you make id a class member of SimpleHtml5Audio? Is there really a need for a separate class? That will save you some more lines.

I noticed you still have different code paths for 1 audio file and for more than 1 audio file:

```
        if 1 < counter:
            str1 = '<audio id="player_{id}" autoplay controls>\n<source src='.format(id = apc.id)
            text = text.replace('<source src=', str1, 1)
            apc.id += 1
        elif 1 == counter:
```

Can't you just add something directly inside the loop just above that works in all case? I.e., replace

```
text += "<source src=\"" + filename + "\">" + '\n'
```
with something like

```
text += '<audio id="player_{id}" autoplay controls>\n<source src='.format(id = self.id) + filename + "\">" + '\n'
```
Wouldn't something like that not also work even if there's only 1 mp3?

In that case, a lot of code could go :-)


You can remove " - not very well suited for multiple audio files."

Also, feel free to add you email to the header of the files you made big changes to.

**Comment by jooseef** [comment]
> I think it's called both for the question and for the answer at the same time

The question data appears twice.

>But I'm not really understanding when you are incrementing the id.

apc.id += 1

>It seems to me that all player ids in the question have id=1, and all those in the answer id=2. Is that what's happening? 

No, the ID is incremented by 1, when a player has to play more than one sound file.

>I would expect that you'd want all the player ids different, even within Q and within A.

Yes, that's happening.

> I noticed you still have different code paths for 1 audio file and for more than 1 audio file:

Yes. A html-player get's an ID only if two ore more files have to be played. Players that have only one file, don't need the JS insertion in the HTML code.

>BTW, why don't you make id a class member of SimpleHtml5Audio? Is there really a need for a separate class? That will save you some more lines.

I need an object that "lives", while the webserver-review is running. It don't need to be a class, a simple global variable would be okay too. 

>You can remove " - not very well suited for multiple audio files."

Yes, maybe also removing "- no autoplay to prevent synchronisation bugs." ?
As  autoplay is currently enabled. Users who don't want autoplay, could disable it in the browser settings.

**Comment by pbienst** [comment]
I'm not doing a good job of bringing my points across, sorry...

1) Yes, the JS is not needed for a single file, but my point is that it should work for a single file as well. So, no need to keep the old scheme. That just leads to more complicated code which is difficult to maintain.

2) Your call to increase id seems to happen after the loop over the different files, not during the loop. Hence my comment that all the files in each answer or each question have the same id. Or am I missing something?

**Comment by jooseef** [comment]
The id is increased after the loop, right. But the files don't get an id, 
only the player gets an id. If there are e.g. 3 files, one.mp3, two.mp3, three.mp3 in an anwer or question,
then a html table row like this will be created:
```
<tr><td><div id="b" class="b">
<audio id="player_1" autoplay controls>
		<source src="one.mp3">
	    <source src="two.mp3">
		<source src="three.mp3">
	 </audio> 
</div></td></tr>
```
The player get's the id "player_1". Not the files.
JavaSript is generated later specifically for this player with id=player_1:
```
var audio_player_1 = null;
let index_1 = {val : 0};
init_player(audio_player_1, 'player_1' , index_1);
```

If an answer or question contains only one file, the player will not get an ID:
```
<tr><td><div id="b" class="b">
<audio autoplay controls>
		<source src="one.mp3">
	 </audio> 
</div></td></tr>
```
and no JavaScript will be generated. For this reason there is the if/elif statement after the for-loop,
player with only one file don't get an id, to optimize the JavaScript generation.
So then, in function to_html in file review_wdgt.py, when an audio player with id=player_1 is found in the html-page, the program knows automatically:
Ayyyy, here I have to create JavaScript for this player. And creates the three lines of JavaScript code shown above:
```
var audio_player_1 = null;
let index_1 = {val : 0};
init_player(audio_player_1, 'player_1' , index_1);
```
Sound players that do not have an id, have to play only one file are not searched for. That's the optimization part.
Assumed also players that have to play only one file would also get an id, the program would have to check each player how
much files it has to play, by counting the source tags. Or these players ids would have to be stored in a list or so. I'd say that would result in more code and cpu load than the tiny small if/elif statement after the for-loop.

And btw also the searching for the player ids(in function to_html()) could be avoided, as all the info passes the for-loop and
could be collected there. Unfortunately the question data passes the loop twice, that confused me and ruined my plan as I din't want to look for the player ids in the html-code(in fuction to_html()) as I already set the ids in the loop itself. *dang*
But I think I have an idea how to get rid of looking for the player ids already, adding one(or more) counter:
passing the for-loop twice->create the JS for the players in the question, passing the loop a third time->create the JS for the answer. No need for looking for the player ids afterwards I guess. This would increase the efficiency of the program code.
I'd need to test it with the several card types, that mnemosyne offers.
Introducing an additional counter is not the yellow of an egg either. But still better, if it works, than looking for the player IDs I guess.
But for now, we can leave it as it is I suggest.
:-)

**Comment by jooseef** [comment]
No, I hadn't timed it, I had no alternative code. I also think it not a bottleneck. I removed the if/elif statement.

**Review by pbienst — APPROVED** [review]
Thanks, I will merge this!

**Comment by jooseef** [comment]
👍

**Pull request merged**